clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -4073,6 +4221,51 @@ LValue CodeGenFunction::EmitArraySubscriptExpr(const ArraySubscriptExpr *E,
       ArrayLV = EmitLValue(Array);
     auto *Idx = EmitIdxAfterBase(/*Promote*/true);
 
+    if (SanOpts.has(SanitizerKind::ArrayBounds)) {
+      // If the array being accessed has a "counted_by" attribute, generate
+      // bounds checking code. The "count" field is at the top level of the
+      // struct or in an anonymous struct, that's also at the top level. Future
+      // expansions may allow the "count" to reside at any place in the struct,
+      // but the value of "counted_by" will be a "simple" path to the count,
+      // i.e. "a.b.count", so we shouldn't need the full force of EmitLValue or
+      // similar to emit the correct GEP.
+      const LangOptions::StrictFlexArraysLevelKind StrictFlexArraysLevel =
+          getLangOpts().getStrictFlexArraysLevel();
+
+      if (const auto *ME = dyn_cast<MemberExpr>(Array);
+          ME &&
+          ME->isFlexibleArrayMemberLike(getContext(), StrictFlexArraysLevel) &&
+          ME->getMemberDecl()->hasAttr<CountedByAttr>()) {
+        RecordDecl *RD = ME->getMemberDecl()
+                             ->getDeclContext()
+                             ->getOuterLexicalRecordContext();
```

Comment:
The first example is why I'm using `getOuterLexicalRecordContext`, because we want to start looking for the count field from the beginning of the most-enclosing struct.

The second example isn't currently allowed (you'll get an error unless the `count` field is in the outer structure). It's probably a bug when sema checking the attribute. I'll need to address that, but I'd like to do it as a separate patch. If the second example is an anonymous struct, then it's iffy if you can define that as a flexible array on its own. If it's a "named" struct, then it should be quite possible to do that.

Reply:
How is the first example supposed to work?  Is the expectation that `__counted_by` will only be enforced if the access expression happens to previously pass through some instance of `A`, and it better be the instance we have in mind?  What happens if I just do `someB->member`?

Reply:
Basically, yes. If you have something like:

```
void foo(struct B *b) {
  bar(b->member[12]);`
}
```

Then there won't be a check, because we won't be able to "see" the `count` in `struct A`. I know it's restrictive, but it's the best we can do given that `someB` may not be allocated within an enclosing `struct A`.